tcljava/doc/Extension.3
'\"
'\" Copyright (c) 1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id: Extension.3,v 1.1 1998/10/14 21:09:09 cvsadmin Exp $
'\" 
.so man.macros
.TH Extension 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.Extension -- The base class for building dynamically loadable Tcl
extensions.

.SH METHODS
.sp
void \fBinit\fR(Interp \fIinterp\fR)
.sp
void \fBsafeInit\fR(Interp \fIsafeInterp\fR)
.sp
static void \fBloadOnDemand\fR(Interp \fIinterp\fR, String
\fIcmdName\fR, String \fIclsName\fR)

.SH ARGUMENTS
.AS TclObject "table[]" .
.AP Interp interp
The interpreter in which the extension is to be loaded.
.AP Interp safeInterp
The interpreter in which the extension is to be loaded safely.
.AP String cmdName
Name of the Tcl command to be loaded.
.AP String clsName
The full pathname of the Java class to load as a Tcl command. The
class should implement the \fBCommand\fR interface.
.BE

.SH DESCRIPTION

The Extension class is used to create dynamic Tcl extension, which can
be loaded into a running Jacl interpreter with the \fBjava::load\fR
Tcl command

.TP
\fBinit\fR
A dynamic Tcl extension should override the \fBinit\fR method to to
perform initialization specific to the given extension. Usually, this
means creating new commands inside the interpreter.
.TP
\fBsafeInit\fR
The \fBsafeInit\fR method is similar to \fBinit\fR, but it's used to
perform initializations when the package is loaded into a safe
interpreter. It should only enable features that are deemed "safe" to
be executed by untrusted Tcl scripts.
.sp
Note: safe interpreters are not currently implemented in Jacl and this
function is currently only for future compatibility.

.TP
\fBloadOnDemand\fR
The \fBloadOnDemand\fR method arranges a \fBCommand\fR implementation
class, specified by \fIclsName\fR, to be loaded into the JVM the first
time the command \fIcmdName\fR is executed; this can reduce the time
it takes to initialize an Extension. The class must implement either
the \fBCommand\fR interface or the \fBCommandWithDispose\fR interface.

.SH "SEE ALSO"
Interp, Command, CommandWithDispose

.SH KEYWORDS
loading, dynamic extension, safe interpreter



tcljava/ChangeLog
2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * README.jacl:
        * README.tclblend: Mention that Jacl and Tcl Blend
        are production ready and include URLs for build
        notes on the wiki.

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        Update version numbers to 1.3.3.

        * configure:
        * configure.in:
        * src/pkgIndex.tcl:
        * src/jacl/tcl/lang/Interp.java:
        * src/tcljava/tcl/lang/BlendExtension.java:
        * unix/tools/vutil/current:
        * win/makefile.vc:

2006-06-29  Mo DeJong  <mdejong@users.sourceforge.net>

        * configure: Regen.
        * configure.in: Set default --prefix to /usr/local.

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/tclblend/tcl/lang/Interp.java (evalFile):
        Add TclBlend implementation of Interp.evalFile().

2006-04-05  Mo DeJong  <mdejong@users.sourceforge.net>

        * src/jacl/tcl/lang/ExecCmd.java (cmdProc):
        Reimplement JDK 1.3 and newer exec so that two
        threads are created. One reads stdout of the
        subprocess while the other reads stderr. This
        is needed to avoid deadlocking in waitFor()
        when a process writes a lot of stdout and stderr.
        * src/tests/tcljava/tests/exec/ExecLotsOutput.java:
        * tests/inprogress/ExecCmd.test: Add test case
        for the waitFor() exec deadlocking bug.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-branch branch and
 	and tcljava-132-branchpoint tags.

2005-08-08  Mo DeJong  <mdejong@users.sourceforge.net>

	Add tcljava-132-2005-08-08 tag to mark tree
	state for the 1.3.2 release.

2005-07-28  Mo DeJong  <mdejong@users.sourceforge.net>

	* announce: Update.
	* docs/website/index.html: Update.

2005-07-27  Mo DeJong  <mdejong@users.sourceforge.net>

	* README.jacl: Update msys description and
	mention jaclsh.bat script.
	* README.tclblend: Mention Tcl 8.4.11 and
	Thread 2.6.2 reqs. Add msys url and
	mention jtclsh.bat script.
	* docs/website/faq.html: Update FAQ.
	* docs/website/index.html: Update homepage
	info with new 1.3.2 release.
	* docs/website/mail.html: Add mail page that
	explains how the mailing lists work.

2005-07-23  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Channel.java (initOutput):
	When initializing a channel of type "file",
	set a flag to indicate that the channel
	should be sync()'ed in addition to flushing.
	* src/jacl/tcl/lang/FlushCmd.java: Print
	IO exception stack trace, this should never
	happen but if it did sa stack trace would
	be helpful.
	* src/jacl/tcl/lang/TclOutputStream.java (setSync,
	flushChannel):
	Add flag setting method so that an output
	stream can be marked so that sync() will
	be invoked when flushing.
	* tests/tcl/SocketCmd.test: Add simple socket
	test case that writes to and reads from a socket.

2005-07-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/BinaryCmd.java (FormatNumber,
	ScanNumber):
	Handle edge cases when casting from a double
	to a float. Rewrite scanning code so that
	64 bit long value is used for shifting operation.
	* src/jacl/tcl/lang/InfoCmd.java:
	* src/jacl/tcl/lang/Shell.java:
	* src/jacl/tcl/lang/SubstCmd.java:
	Invoke static method via class name.
	* src/jacl/tcl/lang/Util.java (strtod): Add
	support for "NaN", "Inf", and "-Inf" strings
 	as valid double values.
	* tests/tcl/binary.test: Enable test cases
	that were not being run because they had
	the nonPortable constraint. Java supports
	only big endian primitive types so the
	binary format and scan operations can
	just assume big endian layout. Test that
	assume little endian are not run via
	addition of the littleEndian constraint.
	Add test cases for double -> float edge cases.

2005-07-20  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Copy new files into dist.
	* tests/tclblend/loading.test: Rework test
	so that it is only run under Win32 and
	returns the proper result.
	* win/jaclsh.bat: Removed old bat script.

2005-07-19  Mo DeJong  <mdejong@users.sourceforge.net>

	* Makefile.in: Remove old green threads code
	for "custom" target. Tcl Blend only works
	with a native threads JVM anyway. Add new
	custom2 shell which loads a JVM at runtime
	and then loads Tcl Blend like a JVM would.
	This shell make it easier to debug weird
	loading issues in Tcl Blend.
	* src/tests/tclblend/tests/JavaEval.java:
	Make the program exit(0) when an expected
	result matches the program result. Use
	positive exit status.
	* tests/tclblend/loading.test: Use the
	expected compare test in the JavaEval
	class.
	* unix/custom2.c: Add new custom shell impl
	that loads a JVM and then loads Tcl Blend into
	the JVM.

2005-07-18  Mo DeJong  <mdejong@users.sourceforge.net>

	Fix crashes when loading Tcl Blend into an
	existing JVM by invoking Tcl_FindExecutable("java")
	to init Tcl subsystems.

	* src/native/javaCmd.c: Update comment.
	* src/native/javaInterp.c (
	Java_tcl_lang_Interp_create,
	Java_tcl_lang_Interp_init,
	Java_tcl_lang_Interp_initName): Add
	TCLBLEND_DEBUG output statements.
	Add initName() method that will init the
	Tcl executable name when Tcl Blend is loaded
	from Java.
	* src/tclblend/tcl/lang/Interp.java (shlib_load,
	initName):
	Add initName() method mentioned above and
	invoke it after loading the Tcl Blend shared lib.
	* tests/tclblend/loading.test: Update tests
	to account for executable name check. Add
	test for [info nameofexecutable] in a Tcl Blend
	interp loaded into Java.

2005-07-17  Mo DeJong  <mdejong@users.sourceforge.net>

	Add startup .bat files for Jacl and Tcl Blend.
	The bat files are auto generated by configure
	with the correct paths.

	* Makefile.in: Install and test jaclsh.bat
	and jtclsh.bat. Make sure the PATH variable
	is always set with the new directories
	prepended to the PATH. Rework tricky gdb
	invocation under Win32 since the mingw
	version of gdb does not know about mingw
	style unix paths.

tcljava/src/tcljava/tcl/lang/Extension.java
/*
 * Extension.java --
 *
 * Copyright (c) 1997 Cornell University.
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: Extension.java,v 1.3 2005/11/04 21:02:14 mdejong Exp $
 *
 */

package tcl.lang;

/**
 * Base class for all Tcl Extensions. A Tcl Extension defines a set of
 * commands that can be loaded into an Interp as a single unit.
 *
 * When a Tcl Extension is loaded into an Interp, either statically
 * (using the "new" operator inside Java code) or dynamically (using
 * the java::load command in Tcl scripts), it usually creates a set of
 * commands inside the interpreter. Occasionally, loading an Extension
 * may lead to additional side effects. For example, a communications
 * Extension may open network connections when it's loaded. Please
 * refer to the documentation of the specific Extension for details.
 */

abstract public class Extension {

    /**
     * Default constructor. Does nothing. The purpose of this
     * constructor is to make sure instances of this Extension can be
     * loaded dynamically using the "java::load" command, which calls
     * Class.newInstance().
     */

    public Extension() {}

    /**
     * Initialize the Extension to run in a normal (unsafe)
     * interpreter. This usually means creating all the commands
     * provided by this class. A particular implementation can arrange
     * the commands to be loaded on-demand using the loadOnDemand()
     * function.
     *
     * @param interp current interpreter.
     */

    abstract public void init(Interp interp) throws TclException;

    /**
     * Initialize the Extension to run in a safe interpreter.  This
     * method should be written carefully, so that it initializes the
     * safe interpreter only with partial functionality provided by
     * the Extension that is safe for use by untrusted code.
     *
     * The default implementation always throws a TclException, so that
     * a subclass of Extension cannot be loaded into a safe interpreter
     * unless it has overridden the safeInit() method.
     *
     * @param safeInterp the safe interpreter in which the Extension should
     *     be initialized.
     */

    public void safeInit(Interp safeInterp) throws TclException {
	throw new TclException(safeInterp, "Extension \"" +
	        getClass().toString() + 
		"\" cannot be loaded into a safe interpreter");
    }

    /**
     * Create a stub command which autoloads the real command the first time
     * the stub command is invoked. Register the stub command in the	
     * interpreter.
     *
     * @param interp current interp.
     * @param cmdName name of the command, e.g., "after".
     * @param clsName name of the Java class that implements this command,
     *     e.g. "tcl.lang.AfterCmd"
     */

    public static final void loadOnDemand(Interp interp, String cmdName,
	    String clsName) {
	interp.createCommand(cmdName, new AutoloadStub(clsName));
    }
}

/**
 * The purpose of AutoloadStub is to load-on-demand the classes that
 * implement Tcl commands. This reduces Jacl start up time and, when
 * running Jacl off a web page, reduces download time significantly.
 */

class AutoloadStub implements Command {
    String className;

    /**
     * Create a stub command which autoloads the real command the first time
     * the stub command is invoked.
     *
     * @param clsName name of the Java class that implements this command,
     *     e.g. "tcl.lang.AfterCmd"
     */
    AutoloadStub(String clsName) {
	className = clsName;
    }

    /**
     * Load the class that implements the given command and execute it.
     *
     * @param interp the current interpreter.
     * @param argv command arguments.
     * @exception TclException if error happens inside the real command proc.
     */
    public void cmdProc(Interp interp, TclObject[] objv) throws TclException {
	Command cmd = load(interp, objv[0].toString());
	cmd.cmdProc(interp, objv);
    }

    /**
     * Load the class that implements the given command, create
     * the command in the interpreter, and return. This helper
     * method is provided so to handle the case where a command
     * wants to create a stub command without executing it.
     * The qname argument should be the fully qualified name
     * of the command.
     */

    Command load(Interp interp, String qname) throws TclException {
	Class cmdClass = null;
	Command cmd;
	try {
	    cmdClass = Class.forName(className);
	} catch (ClassNotFoundException e) {
	    throw new TclException(interp,
		    "ClassNotFoundException for class \"" + className + "\"");
	}

	try {
	    cmd = (Command) cmdClass.newInstance();
	} catch (IllegalAccessException e1) {
	    throw new TclException(interp,
		    "IllegalAccessException for class \"" + cmdClass.getName()
		     + "\"");
	} catch (InstantiationException e2) {
	    throw new TclException(interp,
		    "InstantiationException for class \"" + cmdClass.getName()
		    +
		    "\"");
	} catch (ClassCastException e3) {
	    throw new TclException(interp,
		    "ClassCastException for class \"" + cmdClass.getName()
		    + "\"");
	}
	interp.createCommand(qname, cmd);
	return cmd;
    }
}

